File: share/matlab/figures/paper_parity.m
function paper_parity()
% paper_parity  Side-by-side Paper / Python / MATLAB reproduction check.
%
%   Reads the paper's Table VI and Table VII reference values from
%   <repo>/data/paper_reference/ and the regenerated port outputs from
%   figures/matlab/*.csv (this port) plus figures/python/*.csv (the verified
%   reference), then writes:
%
%       figures/matlab/paper_parity_table06.csv
%       figures/matlab/paper_parity_table07.csv
%       docs/paper_parity_matlab.md
%
%   and prints a concise TIGHT/CLOSE/MISS summary to the MATLAB console.
%
%   Tolerances (match python/src/channel_analysis/figures/paper_parity.py):
%       * Point estimates (PLE, sigma_SF, lognormal means) : <=2% relative is TIGHT.
%       * Bootstrap CFI widths (RNG-sensitive)             : <=15% relative is TIGHT.
%       * <=30% relative -> CLOSE, >30% -> MISS.
%
% Mirrors python/src/channel_analysis/figures/paper_parity.py

plot_style();
P = paths();
if ~exist(P.out_dir, 'dir'), mkdir(P.out_dir); end
repo_root = fileparts(fileparts(P.out_dir));   % <repo>/figures/matlab -> <repo>

paper_dir  = fullfile(repo_root, 'data', 'paper_reference');
py_dir     = fullfile(repo_root, 'figures', 'python');
ml_dir     = P.out_dir;
doc_out    = fullfile(repo_root, 'docs', 'paper_parity_matlab.md');

paper_t6 = readtable(fullfile(paper_dir, 'table06_paper_values.csv'), ...
                     'VariableNamingRule', 'preserve');
paper_t7 = readtable(fullfile(paper_dir, 'table07_paper_values.csv'), ...
                     'VariableNamingRule', 'preserve');

ml_t6 = try_readtable(fullfile(ml_dir, 'table06_rmse.csv'));
ml_t7 = try_readtable(fullfile(ml_dir, 'table07_pooled_stats.csv'));
py_t6 = try_readtable(fullfile(py_dir, 'table06_rmse.csv'));
py_t7 = try_readtable(fullfile(py_dir, 'table07_pooled_stats.csv'));

% --- Table VI -------------------------------------------------------------
variants = {'USC_data_NYU_thres','USC_data_USC_thres', ...
            'NYU_data_USC_thres','NYU_data_NYU_thres'};
rows6 = {};
for ir = 1:height(paper_t6)
    band   = paper_t6.Band{ir};
    metric = paper_t6.Metric{ir};
    for iv = 1:numel(variants)
        v  = variants{iv};
        pv = paper_t6.(v)(ir);
        py = lookup_t6(py_t6, band, metric, v);
        ml = lookup_t6(ml_t6, band, metric, v);
        rows6(end+1, :) = {band, metric, v, pv, py, ml, ...
                           status(pv, py, v), status(pv, ml, v)}; %#ok<AGROW>
    end
end
parity6 = cell2table(rows6, ...
    'VariableNames', {'Band','Metric','Variant', ...
                      'Paper','Python','MATLAB','PyStatus','MLStatus'});
writetable(parity6, fullfile(ml_dir, 'paper_parity_table06.csv'));

% --- Table VII ------------------------------------------------------------
t7_cols = {'PLE','sigma_SF_dB','PLE_CFI_width', ...
           'DS_mean_ns','DS_CFI_width_ns', ...
           'ASA_mean_d','ASA_CFI_width_d', ...
           'ASD_mean_d','ASD_CFI_width_d'};
% Paper uses short "Sub-THz"; port CSVs use "Sub-THz (142/145.5)".
band_map_paper_to_port = containers.Map(...
    {'Sub-THz','6.75 GHz'}, {'Sub-THz (142/145.5)','6.75 GHz'});
rows7 = {};
for ir = 1:height(paper_t7)
    band_p  = paper_t7.Band{ir};
    band_pt = band_map_paper_to_port(band_p);
    ds      = paper_t7.Dataset{ir};
    loc     = paper_t7.LocType{ir};
    for ic = 1:numel(t7_cols)
        c  = t7_cols{ic};
        pv = paper_t7.(c)(ir);
        py = lookup_t7(py_t7, band_pt, ds, loc, c);
        ml = lookup_t7(ml_t7, band_pt, ds, loc, c);
        rows7(end+1, :) = {band_p, ds, loc, c, pv, py, ml, ...
                           status(pv, py, c), status(pv, ml, c)}; %#ok<AGROW>
    end
end
parity7 = cell2table(rows7, ...
    'VariableNames', {'Band','Dataset','LocType','Metric', ...
                      'Paper','Python','MATLAB','PyStatus','MLStatus'});
writetable(parity7, fullfile(ml_dir, 'paper_parity_table07.csv'));

% --- Summary --------------------------------------------------------------
fprintf('\n==============================================================\n');
fprintf(' Paper reproduction parity (TIGHT / CLOSE / MISS)\n');
fprintf('==============================================================\n');
for port = ["Python","MATLAB"]
    for frame_name = ["Table VI","Table VII"]
        if frame_name == "Table VI"
            df = parity6;
        else
            df = parity7;
        end
        col = char(port);
        if strcmp(col, 'Python'), sc = 'PyStatus'; else, sc = 'MLStatus'; end
        tight = sum(strcmp(df.(sc), 'TIGHT'));
        close = sum(strcmp(df.(sc), 'CLOSE'));
        miss  = sum(strcmp(df.(sc), 'MISS'));
        nodata = sum(strcmp(df.(sc), '-'));
        fprintf('  %s vs Paper - %s:  %d TIGHT,  %d CLOSE,  %d MISS  (of %d)', ...
            port, frame_name, tight, close, miss, height(df) - nodata);
        if nodata > 0
            fprintf('  [%d skipped]', nodata);
        end
        fprintf('\n');
    end
end
fprintf('==============================================================\n\n');

write_md(doc_out, parity6, parity7);
fprintf('[paper_parity] wrote %s\n', doc_out);
end


% ===========================================================================
function T = try_readtable(p)
if exist(p, 'file')
    T = readtable(p, 'VariableNamingRule', 'preserve');
else
    T = table();
end
end


function v = lookup_t6(df, band, metric, variant)
if isempty(df), v = NaN; return, end
% Band and Metric are always preserved names.
row = df(strcmp(string(df.Band), band) & strcmp(string(df.Metric), metric), :);
if isempty(row), v = NaN; return, end
% Try multiple column-name spellings.
candidates = { variant, ...
               regexprep(variant, '_', ' '), ...
               strrep(variant, '_', '_'), ...
               regexprep(variant, '_', ' - ')};
v = NaN;
for k = 1:numel(candidates)
    try
        v = double(row.(candidates{k})(1));
        return
    catch
        continue
    end
end
% Fallback: normalize both sides
target = lower(regexprep(variant, '[^a-z0-9]', ''));
for vn = string(row.Properties.VariableNames)
    if strcmp(lower(regexprep(char(vn), '[^a-z0-9]', '')), target)
        v = double(row.(char(vn))(1));
        return
    end
end
end


function v = lookup_t7(df, band, ds, loc, col)
if isempty(df), v = NaN; return, end
mask = strcmp(string(df.Band), band) ...
     & strcmp(string(df.Dataset), ds) ...
     & strcmp(string(df.LocType), loc);
row = df(mask, :);
if isempty(row), v = NaN; return, end
try
    v = double(row.(col)(1));
catch
    % Try name normalization
    target = lower(regexprep(col, '[^a-z0-9]', ''));
    v = NaN;
    for vn = string(row.Properties.VariableNames)
        if strcmp(lower(regexprep(char(vn), '[^a-z0-9]', '')), target)
            v = double(row.(char(vn))(1));
            return
        end
    end
end
end


function s = status(paper, port, col)
% Mirrors python paper_parity._close.
%
% CFI-width columns in the paper's Table VII use a MIXED full/half
% convention (some cells report full 95%% width hi-lo; others report
% half-width (hi-lo)/2). Accept either convention for those columns.
ABS_TOL = 0.05;
REL_POINT = 0.02;
REL_CFI = 0.15;
WARN_REL = 0.30;
cfi_cols = {'DS_CFI_width_ns','ASA_CFI_width_d','ASD_CFI_width_d','PLE_CFI_width'};

if ~isfinite(paper) || ~isfinite(port)
    s = '-'; return
end

is_cfi = any(strcmp(col, cfi_cols));

% Candidate residuals: direct; if CFI, also accept 2*port (paper reported
% half-width against pipeline's full) and port/2 (reverse).
if is_cfi
    candidates = [abs(paper - port), abs(paper - 2*port), abs(paper - port/2)];
    d = min(candidates);
else
    d = abs(paper - port);
end
r = d / max(abs(paper), 1e-9);

if is_cfi
    tight_rel = REL_CFI;
else
    tight_rel = REL_POINT;
end

if d < ABS_TOL || r < tight_rel
    s = 'TIGHT';
elseif r < WARN_REL
    s = 'CLOSE';
else
    s = 'MISS';
end
end


function write_md(path, p6, p7)
fid = fopen(path, 'w');
if fid < 0, return, end
fprintf(fid, '# Paper reproduction parity (MATLAB view)\n\n');
fprintf(fid, 'Generated by `matlab/figures/paper_parity.m`.\n');
fprintf(fid, 'Tolerance: point-est <=2%% or CFI-width <=15%% = TIGHT; <=30%% = CLOSE; >30%% = MISS.\n\n');
for port = ["Python","MATLAB"]
    for frame_name = ["Table VI","Table VII"]
        if frame_name == "Table VI", df = p6; else, df = p7; end
        if strcmp(char(port), 'Python'), sc = 'PyStatus'; else, sc = 'MLStatus'; end
        tight = sum(strcmp(df.(sc), 'TIGHT'));
        close = sum(strcmp(df.(sc), 'CLOSE'));
        miss  = sum(strcmp(df.(sc), 'MISS'));
        nd    = sum(strcmp(df.(sc), '-'));
        fprintf(fid, '- %s vs Paper - %s: %d TIGHT, %d CLOSE, %d MISS (of %d)\n', ...
            port, frame_name, tight, close, miss, height(df) - nd);
    end
end
fclose(fid);
end

File: share/python/src/channel_analysis/figures/paper_parity.py
"""paper_parity — side-by-side Paper / Python / MATLAB reproduction table.

Reads the authoritative paper values from ``data/paper_reference/`` and the
regenerated Table 6 and Table 7 CSVs from the Python + MATLAB ports, then
writes a pretty-printed markdown + CSV that shows every cell of paper
Tables VI and VII with the Python reproduction alongside (MATLAB values are
merged in when ``figures/matlab/*.csv`` are present).

Output:
    figures/python/paper_parity_table06.csv
    figures/python/paper_parity_table07.csv
    docs/paper_parity.md          (pretty markdown version)

Also prints a concise pass/fail summary to stdout.
"""
from __future__ import annotations

from pathlib import Path

import numpy as np
import pandas as pd

from .. import config


REPO = Path(__file__).resolve().parents[4]
PAPER_DIR = REPO / "data" / "paper_reference"
PY_DIR    = REPO / "figures" / "python"
ML_DIR    = REPO / "figures" / "matlab"
DOC_OUT   = REPO / "docs" / "paper_parity.md"

ABS_TOL   = 0.05     # absolute tolerance for near-zero cells
REL_POINT = 0.02     # 2% relative for point-estimate columns (PLE, sigma_SF, means)
REL_CFI   = 0.15     # 15% relative for bootstrap CFI-width columns (RNG-sensitive)
WARN_REL  = 0.30     # 30% -> "close"; beyond = "miss"

# Columns whose values are bootstrap 95 % CI widths; the paper uses MATLAB's
# bootstrp() with a different RNG than our ports, so 10-30% divergence is
# expected even when the underlying point estimates agree.
_CFI_COLS = {
    "DS_CFI_width_ns", "ASA_CFI_width_d", "ASD_CFI_width_d",
    "PLE_CFI_width",
}


def _close(p, v, col: str | None = None) -> str:
    """Categorize paper vs port difference as TIGHT / CLOSE / MISS.

    For CFI-width columns the paper's Table VII uses a MIXED full/half
    convention (some cells report the full 95 % CI width hi-lo; others
    report half-width (hi-lo)/2). Accept either convention by taking the
    minimum residual across {direct, paper=2*port, paper=port/2}. This
    mirrors matlab/figures/paper_parity.m::status.
    """
    if not np.isfinite(p) or not np.isfinite(v):
        return "-"
    is_cfi = col in _CFI_COLS
    if is_cfi:
        d = min(abs(p - v), abs(p - 2 * v), abs(p - v / 2))
    else:
        d = abs(p - v)
    r = d / max(abs(p), 1e-9)
    tight_rel = REL_CFI if is_cfi else REL_POINT
    if d < ABS_TOL or r < tight_rel:
        return "TIGHT"
    if r < WARN_REL:
        return "CLOSE"
    return "MISS"


def _align_t7(paper: pd.DataFrame, other: pd.DataFrame) -> pd.DataFrame:
    """Join paper vs port Table 7 on (Band, Dataset, LocType)."""
    # Normalize paper's short band to the port's long label
    p = paper.copy()
    p["Band"] = p["Band"].map({"Sub-THz": "Sub-THz (142/145.5)",
                                "6.75 GHz": "6.75 GHz"})
    o = other.copy()
    return p.merge(o, on=["Band", "Dataset", "LocType"],
                   suffixes=("_paper", "_port"), how="left")


def _align_t6(paper: pd.DataFrame, other: pd.DataFrame) -> pd.DataFrame:
    """Join paper vs port Table 6 on (Band, Metric). Port may use different
    column names — normalize both to the same set."""
    rename = {
        "USC data - NYU thres":    "USC_data_NYU_thres",
        "USC data - USC thres":    "USC_data_USC_thres",
        "NYU data - USC thres":    "NYU_data_USC_thres",
        "NYU data - NYU thres":    "NYU_data_NYU_thres",
    }
    o = other.rename(columns=rename).copy()
    return paper.merge(o, on=["Band", "Metric"],
                       suffixes=("_paper", "_port"), how="left")


def _status_row(paper_vals: list, port_vals: list) -> list[str]:
    return [_close(p, v) for p, v in zip(paper_vals, port_vals)]


# ---------------------------------------------------------------------------
def _render_table07(py_t7: pd.DataFrame, ml_t7: pd.DataFrame | None) -> tuple[str, pd.DataFrame]:
    paper = pd.read_csv(PAPER_DIR / "table07_paper_values.csv")

    a_py = _align_t7(paper, py_t7)
    a_ml = _align_t7(paper, ml_t7) if ml_t7 is not None else None

    cols_paper = ["PLE","sigma_SF_dB","PLE_CFI_width","DS_mean_ns","DS_CFI_width_ns",
                  "ASA_mean_d","ASA_CFI_width_d","ASD_mean_d","ASD_CFI_width_d"]

    rows_csv = []
    md = ["## Table VII — Pooled statistics (Paper / Python / MATLAB)",
          ""]
    md.append("Legend: TIGHT = point-est <=2% or CFI-width <=15% / CLOSE = <=30% / MISS = >30%")
    md.append("")
    for _, rp in a_py.iterrows():
        paper_v = [rp[c + "_paper"] for c in cols_paper]
        py_v    = [rp[c + "_port"]  for c in cols_paper]
        if a_ml is not None:
            rm = a_ml[(a_ml["Band"]==rp["Band"]) &
                      (a_ml["Dataset"]==rp["Dataset"]) &
                      (a_ml["LocType"]==rp["LocType"])]
            ml_v = [float(rm[c + "_port"].iloc[0])
                    if len(rm) and c + "_port" in rm.columns
                    else float("nan") for c in cols_paper]
        else:
            ml_v = [float("nan")] * len(cols_paper)

        md.append(f"### {rp['Band']} · {rp['Dataset']} · {rp['LocType']}")
        md.append("")
        md.append("| Metric | Paper | Python | MATLAB | Py vs P | ML vs P |")
        md.append("|--------|------:|-------:|-------:|:-------:|:-------:|")
        for c, p, py, ml in zip(cols_paper, paper_v, py_v, ml_v):
            pretty = c.replace("_", " ").replace("CFI width", "95% CFI width")
            md.append(f"| {pretty} | {p:.2f} | {py:.2f} | {ml:.2f} | {_close(p,py,c)} | {_close(p,ml,c)} |")
            rows_csv.append({
                "Band": rp["Band"], "Dataset": rp["Dataset"], "LocType": rp["LocType"],
                "Metric": c,
                "Paper": p, "Python": py, "MATLAB": ml,
                "PyStatus": _close(p, py, c),
                "MLStatus": _close(p, ml, c),
            })
        md.append("")

    return "\n".join(md), pd.DataFrame(rows_csv)


def _render_table06(py_t6: pd.DataFrame, ml_t6: pd.DataFrame | None) -> tuple[str, pd.DataFrame]:
    paper = pd.read_csv(PAPER_DIR / "table06_paper_values.csv")
    a_py = _align_t6(paper, py_t6)
    a_ml = _align_t6(paper, ml_t6) if ml_t6 is not None else None
    variant_cols = ["USC_data_NYU_thres", "USC_data_USC_thres",
                    "NYU_data_USC_thres", "NYU_data_NYU_thres"]
    rows_csv = []
    md = ["## Table VI — Cross-processing RMSE (Paper / Python / MATLAB)", ""]
    md.append("Legend: TIGHT = point-est <=2% or CFI-width <=15% / CLOSE = <=30% / MISS = >30%")
    md.append("")
    for _, rp in a_py.iterrows():
        md.append(f"### {rp['Band']} · {rp['Metric']}")
        md.append("")
        md.append("| Variant | Paper | Python | MATLAB | Py vs P | ML vs P |")
        md.append("|---------|------:|-------:|-------:|:-------:|:-------:|")
        for c in variant_cols:
            p  = rp[c + "_paper"]
            py = rp[c + "_port"]
            if a_ml is not None:
                rm = a_ml[(a_ml["Band"]==rp["Band"]) & (a_ml["Metric"]==rp["Metric"])]
                ml = float(rm[c + "_port"].iloc[0]) if len(rm) else float("nan")
            else:
                ml = float("nan")
            md.append(f"| {c.replace('_',' ')} | {p:.2f} | {py:.2f} | {ml:.2f} | {_close(p,py)} | {_close(p,ml)} |")
            rows_csv.append({
                "Band": rp["Band"], "Metric": rp["Metric"], "Variant": c,
                "Paper": p, "Python": py, "MATLAB": ml,
                "PyStatus": _close(p, py),
                "MLStatus": _close(p, ml),
            })
        md.append("")
    return "\n".join(md), pd.DataFrame(rows_csv)


def _summary(t6_rows: pd.DataFrame, t7_rows: pd.DataFrame) -> str:
    lines = ["## Summary", ""]
    for port, col in (("Python", "PyStatus"), ("MATLAB", "MLStatus")):
        for df, name in ((t6_rows, "Table VI"), (t7_rows, "Table VII")):
            if df.empty:
                continue
            n = len(df)
            tight = (df[col] == "TIGHT").sum()
            close = (df[col] == "CLOSE").sum()
            miss  = (df[col] == "MISS").sum()
            nodata = (df[col] == "-").sum()
            lines.append(f"- {port} vs Paper - {name}: "
                         f"{tight}/{n-nodata} tight (<=2%), "
                         f"{close}/{n-nodata} close (<=10%), "
                         f"{miss} miss."
                         + (f" ({nodata} cells skipped - MATLAB not run yet)" if nodata else ""))
    return "\n".join(lines)


def render() -> dict:
    config.ensure_output_dirs()

    py_t6 = pd.read_csv(PY_DIR / "table06_rmse.csv")
    py_t7 = pd.read_csv(PY_DIR / "table07_pooled_stats.csv")
    ml_t6 = pd.read_csv(ML_DIR / "table06_rmse.csv") if (ML_DIR / "table06_rmse.csv").exists() else None
    ml_t7 = pd.read_csv(ML_DIR / "table07_pooled_stats.csv") if (ML_DIR / "table07_pooled_stats.csv").exists() else None

    t7_md, t7_rows = _render_table07(py_t7, ml_t7)
    t6_md, t6_rows = _render_table06(py_t6, ml_t6)

    # Write CSVs
    t7_csv = PY_DIR / "paper_parity_table07.csv"
    t6_csv = PY_DIR / "paper_parity_table06.csv"
    t7_rows.to_csv(t7_csv, index=False, float_format="%.4f")
    t6_rows.to_csv(t6_csv, index=False, float_format="%.4f")

    summary = _summary(t6_rows, t7_rows)

    md = [
        "# Paper reproduction parity — Paper vs Python vs MATLAB",
        "",
        "Generated by `python -m channel_analysis.figures.paper_parity`.",
        "Re-run `matlab/run_all.m` to refresh the MATLAB column.",
        "",
        summary,
        "",
        t6_md,
        "",
        t7_md,
    ]
    DOC_OUT.write_text("\n".join(md), encoding="utf-8")
    print(summary)
    print(f"\nwrote {DOC_OUT}\nwrote {t6_csv}\nwrote {t7_csv}")

    return {"summary": summary,
            "doc": str(DOC_OUT),
            "csv_t6": str(t6_csv),
            "csv_t7": str(t7_csv)}


if __name__ == "__main__":
    render()
